### PR TITLE
feat(DTFS2-7669): add portal amendments feature flag middleware

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
We will need lots of endpoints behind feature flags for the portal amendments work. Using middleware will keep this dry

## Resolution :heavy_check_mark:
- Add frontend middleware to gef-ui
- Add backend middleware to libs/common (to use in portal-api and dtfs-central-api)
- Add unit test coverage for both

## Miscellaneous :heavy_plus_sign:
- Make deal cancellation middleware more consistent & fix typos

### CODE REVIEW

File: libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Creates a middleware function to check if the given feature flag is enabled
+ *
+ * This will send status `404` if the feature flag is disabled and call `next` otherwise
+ */
+const generateBackendFeatureFlagMiddleware =
+  (isFeatureFlagEnabled: () => boolean): RequestHandler =>
+  (req, res, next) => {
+    if (!isFeatureFlagEnabled()) {
+      console.info(`Feature flag disabled, accessing ${req.originalUrl}`);
```
Comment (ttbarnes) [inline libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts:14]: Is there a way we can consume and log out the feature flag name here? Feels cryptic at the moment. 

```ts
console.info(`Feature flag ${featureFlagName} disabled whilst accessing ${req.originalUrl}`);
```

File: libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Creates a middleware function to check if the given feature flag is enabled
+ *
+ * This will redirect to `/not-found` if the feature flag is disabled and call `next` otherwise
+ */
+export const generateFrontendFeatureFlagMiddleware =
+  (isFeatureFlagEnabled: () => boolean): RequestHandler =>
+  (req, res, next) => {
+    if (!isFeatureFlagEnabled()) {
+      console.info(`Feature flag disabled, accessing ${req.originalUrl}`);
```
Comment (ttbarnes) [inline libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts:30]: Same as above

**Review by ttbarnes — CHANGES_REQUESTED** [review]

File: libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Creates a middleware function to check if the given feature flag is enabled
+ *
+ * This will redirect to `/not-found` if the feature flag is disabled and call `next` otherwise
+ */
+export const generateFrontendFeatureFlagMiddleware =
+  (featureFlagName: FeatureFlag, isFeatureFlagEnabled: () => boolean): RequestHandler =>
+  (req, res, next) => {
+    if (!isFeatureFlagEnabled()) {
+      console.info(`Feature flag ${featureFlagName} disabled, accessing ${req.originalUrl}`);
```
Comment (ttbarnes) [inline libs/common/src/middleware/feature-flags/validate-feature-flag-enabled.ts:31]: Can we add "whilst" or "when" please:

```suggestion
      console.info(`Feature flag ${featureFlagName} disabled whilst accessing ${req.originalUrl}`);
```

**Review by BethThomas141 — APPROVED** [review]
Looks great, nice commonising of middleware 😃 thanks for doing this!

**Review by ttbarnes — APPROVED** [review]

**Pull request merged**